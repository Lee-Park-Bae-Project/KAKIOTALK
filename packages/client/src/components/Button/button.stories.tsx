import React from 'react';
import { withKnobs } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import Button from './index';


export default {
  title: 'Component/Button',
  component: Button,
  decorators: [withKnobs],
  parameters: {
    componentSubtitle: '기본적인 버튼',
  },
};

export const BasicButton: React.FC = () => (
  <Button
    text='primary'
    theme='primary'
  />
);

export const SecondaryButton: React.FC = () => (
  <Button
    text='secondary'
    theme='secondary'
  />
);

export const TertiaryButton = () => (
  <Button
    text='tertiary'
    theme='tertiary'
  />
);

export const NotAllowedButton = () => (
  <Button
    text='전송'
    isAllowd={false}
  />
);

export const ButtonWithLongText: React.FC = () => (
  <Button
    text="Miusov, as a man man of breeding and deilcacy, could not but feel some inwrd qualms, when he reached the Father Superior's with Ivan: he felt ashamed"
    />
);

export const ButtonWithClickEvent: React.FC = () => (
    <Button
      text='click me'
      onClick={action('onClick')}
    />

);
