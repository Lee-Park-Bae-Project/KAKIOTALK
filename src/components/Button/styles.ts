import styled from 'styled-components';
import { darken } from 'polished';
import { color, fontSize, weight } from 'styles/global';

export const themeMap = {
  primary: {
    background: color.YELLO,
  },
  secondary: {
    background: color.RED,
  },
  tertiary: {
    background: color.GRAY,
  },
};
interface ButtonProp{
  isAllowed: boolean;
}

export const Button = styled.button<ButtonProp>`
  width: 5rem;
  height: 2rem;
  color: ${color.BLACK};
  font-size: ${fontSize.SMALL};
  font-weight: ${weight.NORMAL};

  margin: 0.5rem;

  background-color: ${(props) => props.theme.background};

  outline: none;
  border-radius: 1rem;
  border: none;
  cursor: ${(props) => (props.isAllowed ? 'pointer' : 'not-allowed')};

  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;


  &:hover{
    background-color: ${(props) => darken(0.1, props.theme.background)};
  }
`;
