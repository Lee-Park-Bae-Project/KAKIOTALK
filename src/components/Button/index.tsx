import React, { FC } from 'react';
import { ThemeProvider } from 'styled-components';
import * as S from 'components/Button/styles';

interface ButtonProp {
  /** 버튼안에 들어갈 텍스트 */
  text: string;
  /** theme */
  theme?: 'primary' | 'secondary' | 'tertiary';
  /** 사용가능한 버튼인지 */
  isAllowd?: boolean;
  /** 클릭 핸들러 */
  onClick?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}

/**
 *
 *  - `text` 가 너무 길 경우 ... 로 표시되어 나타나지 않습니다.
 * - `onClick`에 props로 핸들러를 전달할 수 있습니다.
 */
const Button: FC<ButtonProp> = ({
  text,
  theme = 'primary',
  isAllowd = true,
  onClick,
}) => (
  <ThemeProvider theme={S.themeMap[theme]}>
    <S.Button
      type='button'
      isAllowed={isAllowd}
      onClick={onClick}
      >
      {text}
    </S.Button>
  </ThemeProvider>

);

export default Button;
