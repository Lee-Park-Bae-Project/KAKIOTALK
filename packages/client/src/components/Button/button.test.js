/* eslint-disable no-undef */
import React from 'react';
import { mount } from 'enzyme';
import Button from 'components/Button/index';

describe('<Button />', () => {
  it('matches snapshot', () => {
    const wrapper = mount(<Button test="button"/>);
    expect(wrapper).toMatchSnapshot();
  });

  it('버튼 text 확인', () => {
    const wrapper = mount(<Button text='text'/>);
    expect(wrapper.props().text).toBe('text');
    const button = wrapper.find('button');
    expect(button.text()).toBe('text');
  });

  it('버튼 배경색 확인', () => {
    const wrapper = mount(<Button text='text'/>);
    const button = wrapper.find(Button);
    expect(getComputedStyle(button.getDOMNode()).getPropertyValue('background-color')).toBe('rgb(247, 207, 70)');
  });
});
