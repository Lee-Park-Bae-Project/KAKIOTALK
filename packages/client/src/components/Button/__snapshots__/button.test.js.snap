// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matches snapshot 1`] = `
<Button
  test="button"
>
  <ThemeProvider
    theme={
      Object {
        "background": "#F7CF46",
      }
    }
  >
    <styled.button
      isAllowed={true}
      type="button"
    >
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": false,
              "lastClassName": "yPZVg",
              "rules": Array [
                "
  width: 5rem;
  height: 2rem;
  color: ",
                "#4B4B4B",
                ";
  font-size: ",
                "0.8rem",
                ";
  font-weight: ",
                "400",
                ";

  margin: 0.5rem;

  background-color: ",
                [Function],
                ";

  outline: none;
  border-radius: 1rem;
  border: none;
  cursor: ",
                [Function],
                ";

  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;


  &:hover{
    background-color: ",
                [Function],
                ";
  }
",
              ],
            },
            "displayName": "styled.button",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "button",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
        isAllowed={true}
        type="button"
      >
        <button
          className="sc-bdVaJa yPZVg"
          type="button"
        />
      </StyledComponent>
    </styled.button>
  </ThemeProvider>
</Button>
`;
